### ISSUE TITLE
AddScreen in TWIN

### ISSUE THREAD

**Issue by dimitrik-fr** [cosmos72/twin#39]
From a long time I wanted to test "AddScreen" feature in TWIN, and here is my feedback:

- unfortunately the name of the Screen is now shown in menu bar, so you never know in which exactly Screen you're working right now..
- seems like there is no hot key implemented to switch to Screen N (or prev./next), only mouse click in the menu, but any way to change ": :" labels to something like "<< Name >>" for ex.? (where Name is the current Screen name)
- and finally, there is the menu option "SendToScreen name" allowing to send the current window to the given Screen "name", however if you try to use it -- it'll will hang the whole TWIN session ! (which is not cool ;-))

how to reproduce it: 

add into .twinrc the following:

```
# add Screen 2
AddScreen 2

# background
Background 2 Black (
   " "
)

# menu
AddToMenu " Window " (
   " Move "        Interactive Move
   " Resize "      Interactive Resize
   " Scroll "      Interactive Scroll
   " Center "      Center
   " Maximize "    Maximize
   " Full Screen " FullScreen
   " Roll Up "     Roll Toggle
   ""              Nop
   " Send to Screen #1 " SendToScreen 1
   " Send to Screen #2 " SendToScreen 2
   ""              Nop
   " Raise/Lower " RaiseLower
   " UnFocus "     Focus Off
   " Next "        _Next
   " WinList... "  WindowList
   ""              Nop
   " Refresh "     Refresh
   " Send Pause "  SyntheticKey Pause
   ""              Nop
   " Close "       Close
   ""              Nop
   " Kill Client " Kill
)
```
start TWIN, open Term, and then from Window menu try to send it to Sceeen 2.

Rgds,
-Dimitri

**Comment by cosmos72** [comment]
Multiple screens is one of the least tested features, I am not surprised it accumulated some bugs in these years.
I will try to reproduce your issue.

**Comment by vanrein** [comment]
Named screens would be great.  They make it easier to attach to a session because names are easier to remember.  I also had a look and it seems that the biggest change is that `/tmp/.Twin:3` could become `/tmp/.Twin:dnsmanage0` or such.

Having multiple screens also allows other neat features (that I find interesting because I'm looking into TWIN for systems management of otherwise headless servers including complex container environments).  Things you could do include

  * Allowing members of a group to access a screen
  * Moving or sharing windows on such a shared screen for their review

I'm pretty fed up with Skype as a way of sharing a desktop, and TWIN could be a wonderful replacement.  That's also a second reason for liking to have a ZMODEM transport built in.  I also believe that the silly things that VMs and RunC do with TTYs could be wonderfully merged into window popups in TWIN (or Screen, or Tmux) in a suitable session or screen.

**Comment by cosmos72** [comment]
@vanrein I think you are talking about **sessions**, not **screens**.
A session is an instance of the server, which listens on one of `/tmp/.Twin:NUMBER`
A screen is somewhat similar to a virtual desktop inside a single session.

In any case, your idea to use names instead of numbers to identify a session should be pretty straightforward to implement. Are you interested in trying to implement it?

**Comment by vanrein** [comment]
> @vanrein I think you are talking about **sessions**, not **screens**.

Yes, of course.

> In any case, your idea to use names instead of numbers to identify a session should be pretty straightforward to implement. Are you interested in trying to implement it?

I am.  Finding the time is the main issue.

I would rather complete the ZMODEM integration first, and then there's the [mulTTY]() idea of session multiplexing with [otherwise ignored ASCII codes](https://github.com/arpa2/multty/blob/master/doc/ASCII.MD) that would add much more value, namely a smooth integration with [container management](https://github.com/arpa2/multty/blob/master/doc/CONTAINER.MD).

@cosmos62 I am wondering if you would be interested in doing some TWIN coding as part of our [open source project](http://internetwide.org/blog/2016/06/24/iwo-phases.html) that involves [many containers](http://internetwide.org/blog/2017/05/03/building-with-blocks.html) and could use a good view on those.  If this would be of interest, we should maybe talk about some interesting options: users, (remote and cross-realm) access control, container integration.

**Comment by vanrein** [comment]
> I would rather complete the ZMODEM integration first, and then there's the [mulTTY]() idea of session multiplexing with [otherwise ignored ASCII codes](https://github.com/arpa2/multty/blob/master/doc/ASCII.MD) that would add much more value, namely a smooth integration with [container management](https://github.com/arpa2/multty/blob/master/doc/CONTAINER.MD).

Oh, and I do hope you don't think I'm being too creative.  TWIN is a very inspiring tool to me!

**Comment by cosmos72** [comment]
> @cosmos62 I am wondering if you would be interested in doing some TWIN coding as part of our [open source project](http://internetwide.org/blog/2016/06/24/iwo-phases.html) that involves [many containers](http://internetwide.org/blog/2017/05/03/building-with-blocks.html) and could use a good view on those. If this would be of interest, we should maybe talk about some interesting options: users, (remote and cross-realm) access control, container integration.

Yes, I am interested in adding new features, especially to better integrate with other tools and/or enable new usage scenarios.

As you said, finding the time is the main issue...

**Comment by vanrein** [comment]
Hi Massimiliano,

> Yes, I am interested in adding new features, especially to better integrate with other tools and/or enable new usage scenarios.

My direct email address is rick@openfortress.nl
Skype is "ofotest", SIP and Jitsi are also possible.

(I am removing this email if it appears on the GitHub board, which I don't want to hijack.  Also it's too personal.)

> As you said, finding the time is the main issue...

Don't know if it helps, but we have been supported bij NLnet for a diversity of ARPA2 projects.
https://nlnet.nl

I would not at all be surprised if TWIN's friendly interface, which IMHO makes a lot of sense on server platforms (and to demo them), would be considered by NLnet to be a welcome addition to the ARPA2 project family.

If you like, we can discuss possible integrations with ARPA2, which I believe would match really well with open ends that TWIN still has.  That could benefit an application that you could submit to NLnet.  (I am the architect and cryptographer behind the overall project named "InternetWide Architecture".)  You could of course take the opportunity of getting funds to fix other open ends, like twdialog and a bundle of issues.  And what ARPA2 would help you add is secure remote access, users/aliases/pseudonyms, groups/roles and all access controlled over an ACL, central or locally constructed.

The sort of thing I usually do is cryptographic and protocol-oriented.  I could help you to a protocol design for the TWIN protocol that could go live.  Pretty standard stuff it would roughly be TLS, then [multiplexed] TTY with SASL authentication.  We have components for TLS and SASL to simplify those tremendously (for programmers and for operators) and I'm experimenting with TTY multiplexing with simple ASCII annotations.


If this sounds useful and interesting, then let's talk and see where we get :-)


Cheers,
 -Rick
